### PR TITLE
feat: modernize with PDM

### LINKED ISSUE #303: modernize nebula-py with PDM
ref: https://github.com/pdm-project/pdm

### PR DESCRIPTION
**Pull request by frostming**
Signed-off-by: Frost Ming <me@frostming.com>

Close #303

### CODE REVIEW

**Review by wey-gu — DISMISSED** [review]
THIS IS CRAZY....

THANKS @frostming 🤦‍♂️

We need a badge in readme later.

**Comment by wey-gu** [comment]
Don't know what to say, thanks @frostming for modernizing NebulaGraph-python, and welcome to the community!

cc @Nicole00 @QingZ11 

I prefer to wait till other two ongoing PR being merged to get this merged :-D.

**Comment by wey-gu** [comment]
@Nicole00 I think we could merge this first before the other ongoing PR?

As either way will involve conflict handling, we may merge this ASAP.

If it's agreed we could ask @frostming to rebase for the last time :D

**Comment by Nicole00** [comment]
It's really an amazing pr!
Thanks @frostming for updating the package manager way for this product, It's a great honor for us to receive your pr.

**Comment by Nicole00** [comment]
> @Nicole00 I think we could merge this first before the other ongoing PR?
> 
> As either way will involve conflict handling, we may merge this ASAP.
> 
> If it's agreed we could ask @frostming to rebase for the last time :D

Sorry I already merged the the other pr, and now there's some conflict for this one. @frostming

**Comment by frostming** [comment]
Never mind, I will resolve it.

**Comment by frostming** [comment]
One concern, I saw the master has changed the minimum python version to `3.9` in the platform, but didn't specify it in `python_requires`, which is the standard metadata for this purpose. I just want to confirm are we **officially** dropping support for Python < 3.9? So I can specify it explicitly as `requires_python = ">=3.9"`, is this change okay for the project?

**Comment by Nicole00** [comment]
> One concern, I saw the master has changed the minimum python version to `3.9` in the platform, but didn't specify it in `python_requires`, which is the standard metadata for this purpose. I just want to confirm are we **officially** dropping support for Python < 3.9? So I can specify it explicitly as `requires_python = ">=3.9"`, is this change okay for the project?
> 
> @Nicole00 @wey-gu

Yeah, It's OK to specify it explicitly as `requires_python = ">=3.9"`.

**Comment by Nicole00** [comment]
> One concern, I saw the master has changed the minimum python version to `3.9` in the platform, but didn't specify it in `python_requires`, which is the standard metadata for this purpose. I just want to confirm are we **officially** dropping support for Python < 3.9? So I can specify it explicitly as `requires_python = ">=3.9"`, is this change okay for the project?
> 
> @Nicole00 @wey-gu

Confirm with @wey-gu , we can support >=3.6.  It will be kind if you can help to modify with `requires_python = ">=3.6`.

**Comment by wey-gu** [comment]
Yes, sorry for the confusion, the reason that PR 3.6~~3.8 was dropped was due to the limitation of pip-complier(only a case in dev requirement), now, with the great PDM it's easy to support them all as our production requirements list is quite small.

Please let's keep 3.6+ back if possible with the magic of PDM and Ming.

**Comment by frostming** [comment]
It depends on the compatibility strategy of your project. From a view at January 2024, I would recommend `>=3.8`. But either way is fine for me.

File: .github/workflows/run_test.yaml
Comment (wey-gu) [inline .github/workflows/run_test.yaml:27]: @Nicole00 we could add 3.6 3.7 3.8 back in separate PR after this being merged. :D

**Review by wey-gu — APPROVED** [review]

**Comment by wey-gu** [comment]
> It depends on the compatibility strategy of your project. From a view at January 2024, I would recommend `>=3.8`. But either way is fine for me.

Thanks Ming!

We could consider `>=3.8` in the next major version(4.x). It was PDM that made it possible to not break existing 3.6/7 users in this next release.

File: .github/workflows/run_test.yaml
Comment (frostming) [reply .github/workflows/run_test.yaml:27]: AFAIK the `setup-python` action can no longer install 3.6 **easily**. And some of our dev dependencies don't work on 3.6. https://github.com/vesoft-inc/nebula-python/blob/2eeaef4f5daee904d250945f69c00cc4c44df043/pyproject.toml#L32-L42

Comment (wey-gu) [reply .github/workflows/run_test.yaml:27]: Ah, got it, let's put CI with 3.9+ as is, and keep the requirement broader to defer the decision of changing minimal py.

We could make CI best effort in a slightly broader in other PR.

@Nicole00 This seems to be a good enough timing for at least dropping 3.6 :D

**Comment by wey-gu** [comment]
> 
```
/home/runner/.local/share/pdm/venv/lib/python3.8/site-packages/pdm/resolver/providers.py:158: 
  PackageWarning: Skipping oldest-supported-numpy@2023.12.21 because it requires Python>=3.7 but the project claims to work with Python>=3.6. Instead, another version of oldest-supported-numpy that supports Python>=3.6 will be used.
If you want to install oldest-supported-numpy@2023.12.21, narrow down the `requires-python` range to include this version. For example, ">=3.7" should work.
```

If needed, we could go with `>=3.7` even in this PR.

**Comment by frostming** [comment]
> If needed, we could go with `>=3.7` even in this PR.

OK, let's see if the CI still fails(likely 😢 )

File: pyproject.toml
Comment (wey-gu) [inline pyproject.toml:37]: Maybe we have to give a version for pytest-cov?

ref: https://github.com/nedbat/coveragepy/issues/1376

Comment (frostming) [reply pyproject.toml:37]: Upgraded the version

**Review by wey-gu — APPROVED** [review]

**Review by Nicole00 — APPROVED** [review]

**Comment by wey-gu** [comment]
🥳🎉 cc @QingZ11

**Pull request merged**
